flang/lib/Lower/Bridge.cpp

```diff
@@ -1581,6 +1581,35 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         curEval = &*std::next(curEval->getNestedEvaluations().begin());
       }
     }
+    // If a 'sections' construct is encountered, loop over all 'section'
+    // directives nested within and generate FIR for the same. The individual
+    // 'section' inside a 'sections' construct are not constructs rather
+    // 'Section Blocks'. Hence, nested handling of constructs doesn't apply and
+    // we need special handling
+    if (const auto &ompSections =
+            std::get_if<Fortran::parser::OpenMPSectionsConstruct>(&omp.u)) {
+      const auto &ompSectionBlocks =
+          std::get<Fortran::parser::OmpSectionBlocks>(ompSections->t);
+      auto &sectionsBlockEvaluationList = curEval->getNestedEvaluations();
+      std::list<Fortran::lower::pft::Evaluation>::iterator
+          sectionsBlockEvalIterator = sectionsBlockEvaluationList.begin();
+      for (const auto &block : ompSectionBlocks.v) {
+        auto insertPt = builder->saveInsertionPoint();
+        // create a 'section' operation for every 'Section Block'
+        genOpenMPSectionsBlock(*this, *curEval);
+        for (auto it = block.begin(); it != block.end(); it++) { 
+          // generate FIR for every 'ExecutionPartConstruct' and encapsulate it
+          // within the corresponding 'Section Block'
+          genFIR(*sectionsBlockEvalIterator);
+          sectionsBlockEvalIterator++;
+        }
+        builder->restoreInsertionPoint(insertPt);
+      }
```

Comment:
```suggestion
      for (auto &block : curEval->getNestedEvaluations()) {
        auto insertPt = builder->saveInsertionPoint();
        // create a 'section' operation for every 'Section Block'
        genOpenMPSectionsBlock(*this, *curEval);
        genFIR(block);
        builder->restoreInsertionPoint(insertPt);
      }
```

Reply:
Hello @shraiysh. This is problematic for the reason that it creates a new `section` block for each evaluation under a `sections`. Consider the following: 

```
!$omp sections   
         !$omp section
             count = 1 + 4
             double_count = count + 10
 !$omp end sections
```

And consider the emitted FIR

```
omp.sections  {
    omp.section  {
      %c5_i32 = arith.constant 5 : i32
      fir.store %c5_i32 to %0 : !fir.ref<i32>
      omp.terminator
    }
    omp.section  {
      %2 = fir.load %0 : !fir.ref<i32>
      %c10_i32 = arith.constant 10 : i32
      %3 = arith.addi %2, %c10_i32 : i32
      fir.store %3 to %1 : !fir.ref<i32>
      omp.terminator
    }
    omp.terminator
  }
  return
}
```

Reply:
@shraiysh It is [this](https://github.com/flang-compiler/f18-llvm-project/pull/1289#discussion_r766711889) problem technically.

Reply:
Hmm, the use of long iterator types is unfortunate here and the testcases don't check for this (even with the additional sections, it matches against the other statements.) If possible, please add some -NOT or -NEXT checking in test to capture this. I am OK with this if it isn't feasible to test it with LIT.

Reply:
Please add the suggested test in a later patch.